src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -1050,7 +1050,211 @@ private function step_before_head(): bool {
 	 * @return bool Whether an element was found.
 	 */
 	private function step_in_head(): bool {
-		$this->bail( 'No support for parsing in the ' . WP_HTML_Processor_State::INSERTION_MODE_IN_HEAD . ' state.' );
+		$token_name = $this->get_token_name();
+		$token_type = $this->get_token_type();
+		$is_closer  = parent::is_tag_closer();
+		$op_sigil   = '#tag' === $token_type ? ( $is_closer ? '-' : '+' ) : '';
+		$op         = "{$op_sigil}{$token_name}";
+
+		/*
+		 * > A character token that is one of U+0009 CHARACTER TABULATION,
+		 * > U+000A LINE FEED (LF), U+000C FORM FEED (FF),
+		 * > U+000D CARRIAGE RETURN (CR), or U+0020 SPACE
+		 */
+		if ( '#text' === $op ) {
+			$text = $this->get_modifiable_text();
+			if ( '' === $text ) {
+				/*
+				 * If the text is empty after processing HTML entities and stripping
+				 * U+0000 NULL bytes then ignore the token.
+				 */
+				return $this->step();
+			}
+
+			if ( strlen( $text ) === strspn( $text, " \t\n\f\r" ) ) {
+				// Insert the character.
+				$this->insert_html_element( $this->state->current_token );
+				return true;
+			}
+		}
+
+		switch ( $op ) {
+			/*
+			 * > A comment token
+			 */
+			case '#comment':
+			case '#funky-comment':
+			case '#presumptuous-tag':
+				$this->insert_html_element( $this->state->current_token );
+				return true;
+
+			/*
+			 * > A DOCTYPE token
+			 */
+			case 'html':
+				// Parse error: ignore the token.
+				return $this->step();
+
+			/*
+			 * > A start tag whose tag name is "html"
+			 */
+			case '+HTML':
+				return $this->step_in_body();
+
+			/*
+			 * > A start tag whose tag name is one of: "base", "basefont", "bgsound", "link"
+			 */
+			case '+BASE':
+			case '+BASEFONT':
+			case '+BGSOUND':
+			case '+LINK':
+				$this->insert_html_element( $this->state->current_token );
+				return true;
+
+			/*
+			 * > A start tag whose tag name is "meta"
+			 */
+			case '+META':
+				$this->insert_html_element( $this->state->current_token );
+
+				/*
+				 * > If the active speculative HTML parser is null, then:
+				 * >   - If the element has a charset attribute, and getting an encoding from
+				 * >     its value results in an encoding, and the confidence is currently
+				 * >     tentative, then change the encoding to the resulting encoding.
+				 */
+				$charset = $this->get_attribute( 'charset' );
+				if ( is_string( $charset ) ) {
```

Comment:
Shouldn't this only bail if `$charset` is not UTF-8?

```suggestion
				if ( is_string( $charset ) && ! is_utf8_charset( $charset ) ) {
```

Reply:
yeah there's an interplay here with #6977 that's an artifact due to sharing code but not merging everything at once. the test here is whether we support inferring the character set from a META tag, and that only happens if the parser's charset confidence is `tentative`, but I didn't want to add this code in this PR.

alternatively, we could remove `META` tag support, but I hope that #6977 merges within the next couple of days, making this all moot